### ISSUE TITLE
[JENKINS-37150] Add Whitelist to "Cross Site Request Forgery"

### ISSUE THREAD

**Issue by jenkins-infra-bot** [timja-org/jenkins-gh-issues-poc-11-23#6070]
<p>A number of source-code hosting services (Gitlab, Github, Bitbucket, Gogs, and others) have the ability to send webhooks when new commits are pushed to a repository. The webhooks are usually HTTP POST requests with a JSON payload.</p>

<p>Out of the box, Jenkins 2.x enables a security feature called <em>"Prevent Cross Site Request Forgery exploits"</em> that rejects all POST requests that don't have a valid 'crumb' token associated with them. This is a great security feature.</p>

<p>Since most Git hosting services don't have any mechanism to retrieve or handle a crumb, that means that their POST requests get blocked on a Jenkins installation with default settings. <em>"Prevent Cross Site Request Forgery exploits"</em> needs to be manually disabled in order to allow the host's webhooks to work with Jenkins out-of-the-box, or a service-specific plugin needs to be installed.</p>

<p>There are plugins written for a lot of big-name hosts (e.g. Bitbucket and Github) that provide CrumbExclusions for those specific hosts. This method works (and lets us keep CSRF enabled), but it means that every hosting service needs its own special Jenkins plugin just to trigger a remote SCM poll. It also means that a lot of smaller or self-hosted services get left in the dust.</p>

<p>Would it be possible to add a configurable domain-name whitelist to the CSRF protection feature? That would let a Jenkins installation receive POST requests from specific servers and still keep general CSRF protection in-place. This could maybe take the form of a text-box on the <em>"Configure Global Security"</em> page where we could add whitelisted domains.</p>

<p>This is a general-purpose fix that would permanently solve all issues of this general type:<br/>
<a href="https://issues.jenkins-ci.org/browse/JENKINS-33878" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-33878</a><br/>
<a href="https://issues.jenkins-ci.org/browse/JENKINS-30384" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-30384</a><br/>
<a href="https://issues.jenkins-ci.org/browse/JENKINS-25421" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-25421</a><br/>
<a href="https://issues.jenkins-ci.org/browse/JENKINS-22474" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-22474</a><br/>
<a href="https://issues.jenkins-ci.org/browse/JENKINS-20140" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-20140</a><br/>
<a href="https://issues.jenkins-ci.org/browse/JENKINS-10263" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-10263</a><br/>
<a href="https://issues.jenkins-ci.org/browse/JENKINS-14372" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-14372</a></p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/nrclark.svg" title="nrclark's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nrclark">nrclark</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37150" target="_blank">Add Whitelist to "Cross Site Request Forgery"</a></i></summary>
<i><ul>
<li><b>status</b>: Resolved
<li><b>priority</b>: Minor
<li><b>component(s)</b>: core
<li><b>label(s)</b>: CSRF, Core, Feature, Improvement, Security, UX
<li><b>resolution</b>: Won't Fix
<li><b>resolved</b>: 2016-09-29T06:41:14+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 0
<li><b>imported</b>: 2025-11-23
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>A number of source-code hosting services (Gitlab, Github, Bitbucket, Gogs, and others) have the ability to send webhooks when new commits are pushed to a repository. The webhooks are usually HTTP POST requests with a JSON payload.</p>

<p>Out of the box, Jenkins 2.x enables a security feature called <em>"Prevent Cross Site Request Forgery exploits"</em> that rejects all POST requests that don't have a valid 'crumb' token associated with them. This is a great security feature.</p>

<p>Since most Git hosting services don't have any mechanism to retrieve or handle a crumb, that means that their POST requests get blocked on a Jenkins installation with default settings. <em>"Prevent Cross Site Request Forgery exploits"</em> needs to be manually disabled in order to allow the host's webhooks to work with Jenkins out-of-the-box, or a service-specific plugin needs to be installed.</p>

<p>There are plugins written for a lot of big-name hosts (e.g. Bitbucket and Github) that provide CrumbExclusions for those specific hosts. This method works (and lets us keep CSRF enabled), but it means that every hosting service needs its own special Jenkins plugin just to trigger a remote SCM poll. It also means that a lot of smaller or self-hosted services get left in the dust.</p>

<p>Would it be possible to add a configurable domain-name whitelist to the CSRF protection feature? That would let a Jenkins installation receive POST requests from specific servers and still keep general CSRF protection in-place. This could maybe take the form of a text-box on the <em>"Configure Global Security"</em> page where we could add whitelisted domains.</p>

<p>This is a general-purpose fix that would permanently solve all issues of this general type:
<a href="https://issues.jenkins-ci.org/browse/JENKINS-33878" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-33878</a>
<a href="https://issues.jenkins-ci.org/browse/JENKINS-30384" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-30384</a>
<a href="https://issues.jenkins-ci.org/browse/JENKINS-25421" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-25421</a>
<a href="https://issues.jenkins-ci.org/browse/JENKINS-22474" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-22474</a>
<a href="https://issues.jenkins-ci.org/browse/JENKINS-20140" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-20140</a>
<a href="https://issues.jenkins-ci.org/browse/JENKINS-10263" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-10263</a>
<a href="https://issues.jenkins-ci.org/browse/JENKINS-14372" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-14372</a></p>

</pre>
</details>
</details>
<ul><li><i>environment</i>: <code>Jenkins 2.16</code></li></ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/danielbeck.png" title="danielbeck's avatar" /> danielbeck:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37150?focusedId=271726&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-271726">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Specific endpoints in Jenkins can opt out of requiring CSRF crumbs. That's the approach that should be taken here, rather than burden the admin with something they're likely ill-equipped to deal with.</p>
</pre>
</details>

</ul>
</details>

<p>Specific endpoints in Jenkins can opt out of requiring CSRF crumbs. That's the approach that should be taken here, rather than burden the admin with something they're likely ill-equipped to deal with.</p>
